contracts/0.8.25/vaults/VaultHubViewerV1.sol

```diff
@@ -0,0 +1,137 @@
+
+
+interface IDashboard {
+    function getRoleMember(bytes32 role, uint256 index) external view returns (address);
+    function hasRole(bytes32 role, address account) external view returns (bool);
+}
+
+interface IVault is IStakingVault {
+    function owner() external view returns (address);
+}
+
+interface IVaultHub {
+    function vaultsCount() external view returns (uint256);
+    function vault(uint256 _index) external view returns (IVault);
+}
+
+contract VaultHubViewerV1 {
+    IVaultHub public immutable vaultHub;
+    bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+
+    constructor(address _vaultHubAddress) {
+        if (_vaultHubAddress == address(0)) revert ZeroArgument("_vaultHubAddress");
+        vaultHub = IVaultHub(_vaultHubAddress);
+    }
+
+    function isContract(address account) public view returns (bool) {
+        uint256 size;
+        assembly {
+            size := extcodesize(account)
+        }
+        return size > 0;
+    }
+
+
+    /// @notice Checks if a given address is the owner of a vault
+    /// @param vault The vault to check
+    /// @param _owner The address to check
+    /// @return True if the address is the owner, false otherwise
+    function isOwner(IVault vault, address _owner) public view returns (bool) {
+        address currentOwner = vault.owner();
+        if (currentOwner == _owner) {
+            return true;
+        }
+        if (isContract(currentOwner)) {
+            try IDashboard(currentOwner).hasRole(DEFAULT_ADMIN_ROLE, _owner) returns (bool hasRole) {
+                return hasRole;
+            } catch {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    /// @notice Checks if a given address has a given role on a vault
+    /// @param vault The vault to check
+    /// @param _member The address to check
+    /// @param _role The role to check
+    /// @return True if the address has the role, false otherwise
+    function isHasRole(IVault vault, address _member, bytes32 _role) public view returns (bool) {
+        address owner = vault.owner();
+        if (owner == address(0)) {
+            return false;
+        }
+
+        try IDashboard(owner).hasRole(_role, _member) returns (bool hasRole) {
+                return hasRole;
+        } catch {
+                return false;
+        }
+    }
+
+    /// @notice Returns all vaults owned by a given address
+    /// @param _owner Address of the owner
+    /// @return An array of vaults owned by the given address
+    function vaultsByOwner(address _owner) public view returns (IVault[] memory) {
+        uint256 count = vaultHub.vaultsCount();
+        IVault[] memory vaults = new IVault[](count);
+
+        // Populate the array with the owner's vaults
+        for (uint256 i = 0; i < count; i++) {
+            if (isOwner(vaultHub.vault(i), _owner)) {
+                vaults[i] = vaultHub.vault(i);
+            }
+        }
```

Comment:
```suggestion
        uint256 valid = 0;
        
        // Populate the array with the owner's vaults
        for (uint256 i = 0; i < count; i++) {
            IVault vaultInstance = vaultHub.vault(i);
            if (isOwner(vaultInstance, _owner)) {
                vaults[valid] = vaultHub.vault(i);
                valid++;
            }
        }
```
Added caching and count of nonzero vaults to simplify the filtering

Reply:
fixed applied